#### SC-MSI Snakemake pipeline for use with FASTQ/BAM files ####

# Example command to run #
# snakemake -s handle_fastq.snake --cores 1 --use-conda 

#### BEGIN WORKFLOW ####

#import libraries
import pandas as pd

#define config file
configfile: "fastq_config.yaml"

with open(config["sample_list"]) as samples_tmp:
	samples=samples_tmp.read().splitlines()

#read in patient list
with open(config["patient_list"]) as patient_id:
	patients=patient_id.read().splitlines()

#read in patient-sample key
patient_key=pd.read_table(config["patient_key"])


rule all:
	input:
		expand("../images/{sample_name}_cell_types.pdf",sample_name=samples)
#		expand("../summary_stats/{patient}_cluster_stats.tsv",patient=patients),
#		expand("../patient_reports/{patient}.html",patient=patients)

rule cellranger:
	input:
	output:
		raw_bam="../cell_ranger_output/{sample_name}/outs/possorted_genome_bam.bam",
		matrix="../cell_ranger_output/{sample_name}/outs/filtered_feature_bc_matrix.h5"
	params:
		cellranger=config["cellranger"],
		ref=config["transcriptome"],
		final_dir="../cell_ranger_output/{sample_name}/"
	conda:
		 "base"
	shell:
		"""
		if [[ -d {params.temp_dir} ]]
		then
		rm -r {params.temp_dir}
		fi
		if [[ -d {params.final_dir} ]]
		then
		rm -r {params.final_dir}
		fi

		mkdir -p {params.temp_dir}

		{params.cellranger} count --id={wildcards.sample_name} --transcriptome={params.ref} --fastqs=../fastq/ \
		--sample={wildcards.sample_name} --output-dir {params.final_dir} \
		--localcores {threads}
		"""

rule process_sample:
	input:
		{rules.cellranger.output.matrix}
	output:
		all_barcodes="../pseudobulk_barcodes/{sample_name}/{sample_name}_all_cell_barcodes.tsv",
		just_barcodes="../pseudobulk_barcodes/{sample_name}/just_barcodes.tsv",
		filtered_h5="../filtered_h5/{sample_name}.rds"
	params:
		key=config["patient_key"],
		mtx='N',
		seed=0
	conda:
		"seurat"
	shell:
		"""
		Rscript barcode_generator.R {wildcards.sample_name} {params.key} {params.mtx} {params.seed}
		if [[ ! -f ../pseudobulk_barcodes/{wildcards.sample_name}/just_barcodes.tsv ]]
		then
		cut -f 1 ../pseudobulk_barcodes/{wildcards.sample_name}/{wildcards.sample_name}_all_cell_barcodes.tsv > \
		../pseudobulk_barcodes/{wildcards.sample_name}/just_barcodes.tsv
		fi
		"""

rule find_cancer:
	input:
		"../filtered_h5/{sample_name}.rds"
	output:
		"../pseudobulk_barcodes/{sample_name}/{sample_name}_cancer_barcodes.tsv",
		"../atomic/{sample_name}.rds"
	params:
		seed=0,
	conda:
		"atomic"
	shell:
		"""
		Rscript atomic.R {wildcards.sample_name} {threads} {params.seed}
		"""

rule sensor_rna:
	input:
		"../filtered_h5/{sample_name}.rds"
	output:
		"../sensor_rna_results/{sample_name}.txt"
	conda:
		"seurat.yml"
	shell:
		"""
		Rscript sensor_rna_shaper.R {wildcards.sample_name}
		training_model=../references/{wildcards.sample_name}_training_model.csv
		msisensor-rna train -i $training_model -m ../references/{wildcards.sample_name}.model -t PanCancer
		msisensor-rna detection -i ../references/{wildcards.sample_name}.csv -o ../sensor_rna_results/{wildcards.sample_name}.txt \
		-m ../temp/{wildcards.sample_name}.model -d True
		"""

def get_list_of_bams(wildcards):
	ck_output = checkpoints.split_bam.get(sample_name=wildcards.sample_name).output[0]
	with open(ck_output) as f:
		bams = [line.strip().replace(".bam","") for line in f]
	return bams


checkpoint annotate_h5:
	input:
		"../atomic/{sample_name}.rds",
		"../sensor_rna_results/{sample_name}.txt"
	output:
		"../annotated_h5/{sample_name}.rds"
	conda:
		"seurat"
	shell:
		"""
		Rscript annotate_bamog.R {wildcards.sample_name}
		"""
rule create_sample_plots:
	input:
		{checkpoint.annotate_h5.output}
	output:
		"../images/{sample_name}_cell_types.pdf",
		"../images/{sample_name}_msi_plots.pdf"
	conda:
		"seurat"
	shell:
		"""
		Rscript plot_all_bamog.R {wildcards.sample_name}
		"""

